ramips: add support for ipTIME A104ns & A1004ns

mans0n opened the pull request:
This adds support for ipTIME A104ns and A1004ns with MediaTek MT7620A.

mans0n committed:
ramips: add support for ipTIME A1004ns

ipTIME A1004ns is a 2.4/5GHz band AC750 router, based on MediaTek MT7620A.

Specifications:
- SoC: MT7620A
- RAM: DDR2 128MB
- Flash: SPI NOR 16MB
- WiFi:
  - 2.4GHz: SoC internal
  - 5GHz: MT7610EN
- Ethernet: 5x 10/100/1000Mbps
  - Switch: MT7530BU
- USB: 1x 2.0
- UART:
  - J2: 3.3V, TX, RX, GND (3.3V is the square pad) / 57600 8N1

Installation via web interface:
1.  Flash **initramfs** image through the stock web interface.
2.  Boot into OpenWrt and perform sysupgrade with sysupgrade image.

Revert to stock firmware:
1.  Perform sysupgrade with stock image.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
The ethernet problem can be fixed if `ralink_switchdev` is set properly in `07_set_preinit_iface_ramips`.
Is it a good solution using `board_name` there?

adrianschmutzler commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -641,6 +642,7 @@ ramips_setup_macs()
 		wan_mac=$(mtd_get_mac_binary devdata 0x7)
 		;;
 	edimax,br-6478ac-v2|\
+	iptime,a104ns|\
```
So, nothing on flash for the address?

mans0n replied:
Unlike A1004ns, A104ns has just garbage value at 0x1fc40. How weird.

mans0n commented:
Added preinit iface patch.
I felt like it would be a little bit ugly, but it looks better than I thought.

mans0n commented:
Rebased to remove debounce-interval.

mans0n commented:
Rebased due to 19724e2.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_iptime.dtsi:
```diff
@@ -0,0 +1,97 @@
+
+&gpio0 {
+	status = "okay";
+};
```
gpio0 should be enabled by default, so this node should be obsolete.

mans0n replied:
You're right. Removed.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_iptime.dtsi:
```diff
@@ -0,0 +1,97 @@
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&uboot 0x1f400>;
```
mediatek,mtd-eeprom?

mans0n replied:
mt7620a wmac uses kmod-rt2800 instead of kmod-mt76, and all the other mt7620a dts files use ralink,mtd-eeprom too. I haven't tested mediatek,mtd-eeprom though.

adrianschmutzler replied:
Okay, I won't force you to be the first, unless you are curious.

mans0n replied:
I've just tested with mediatek,mtd-eeprom and it did not work.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_iptime_a104ns.dts:
```diff
@@ -0,0 +1,52 @@
+		led_cpu: cpu {
+			label = "a104ns:blue:cpu";
+			gpios = <&gpio1 15 GPIO_ACTIVE_LOW>;
+		};
```
Really? They just use the inverted gpio<->led assignment compared to the A1004ns?

mans0n replied:
Yes, I can confirm this, and even vendor GPL source code proves it. How weird.
https://github.com/iptime-gpl/userapps_a104ns/blob/master/fs/mt7621/clones/a104ns/si/hw#L24
https://github.com/iptime-gpl/userapps_a1004ns/blob/master/fs/mt7621/clones/a1004ns/si/hw#L25

mans0n commented:
Added a comment in preinit switch script.

mans0n commented:
@981213 Could you please have a look at this?

mans0n commented:
Removed "/dts-v1/;" in dtsi (due to 231dc26867b4)

mans0n commented:
Rebased due to 98d9158d2f67.

adrianschmutzler commented:
I've merged A104ns (with some tiny adjustments) into my staging tree.

Please add a comment about the switched LED GPIOs to A1004ns, so this is preserved for the future.

adrianschmutzler commented:
@mans0n We had a brief discussion on IRC about the preinit_iface changes.
I asked blogic to have a quick look and he stated it looks fine.
ynezz then suggested to move the different "sources" for switchdev into functions and I've written a quick patch: https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

If you have no objections, I'd propose to convert the patch for A1004ns into two steps, first only providing device support (so, just everything without touching the preinit file) and then "my" patch as second. I'd be glad if you could test this combination then. If you prefer, you can obviously also test on the Netis instead (which might be easier due to the half-merged A104ns at the moment).

Thanks for your work so far!

mans0n commented:
@adrianschmutzler It looks much better. Thanks for the new patch. A1004ns works well with it.

mans0n commented:
Rebased due to 38bdfcdb87 and f45a16dead.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -494,6 +494,16 @@ define Device/iodata_wn-ac733gr3
 TARGET_DEVICES += iodata_wn-ac733gr3
 
+define Device/iptime_a1004ns
+  MTK_SOC := mt7620a
```
Rename to "SOC" due to 58e71e322bc0e0edd08157075e9022c15ae33ec5.

mans0n replied:
Renamed.

mans0n commented:
Rebased due to 58e71e3.

mans0n commented:
Rebased due to e417ff88f19b.

mans0n commented:
@adrianschmutzler Does this need another reviewer?
I don't feel sending to PW will help much considering the status of latest patches...

adrianschmutzler commented:
This needs another reviewer as I'm not familiar with this kind of network setup.
I also don't think PW will help here (and you won't get friends by submitting stuff twice), I fear you just have to wait. :-(

mans0n commented:
@adrianschmutzler Okay, so I'd better seek reviewers here.

FWIW I had copied the ethernet settings from mt7620a_iodata_wn-ac1167gr.dts. It seems wn-ac1167gr has switch setup similar to that of a1004ns.
@musashino205 Could you let me know if wn-ac1167gr or wn-ac733gr3 have ethernet problem in failsafe mode or not? (if you still have time and access to the device)

Ref: https://github.com/openwrt/openwrt/commit/c80f67c1985aa1fac7043b5d137ac9eed390b77e
> Known issues:
> - Becomes unreachable via ethernet when in failsafe mode

981213 commented:
If ethernet needs special setup like VLAN, you should use a preinit script to set it up for failsafe. e.g.: target/linux/ramips/base-files/lib/preinit/07_set_preinit_iface_ramips

981213 commented on target/linux/ramips/dts/mt7620a_iptime_a1004ns.dts:
```diff
@@ -0,0 +1,104 @@
+		ethernet-phy@1f {
+			reg = <0x1f>;
+			phy-mode = "rgmii";
+		};
```
Do you actually need all these nodes? It seems that our eth driver will set up external mt7530 automatically with only the 0x1f node.
Could you test if it works without ethernet-phy@0~ethernet-phy@4 nodes?

mans0n replied:
I tried removing phy0~phy4 and port 0 (wan port) stopped working. I needed (at least) phy0 to make it work.
I'll try to investigate this more.

mans0n replied:
I tried removing phy0~phy4 and port 0 (wan port) stopped working. I needed (at least) phy0 to make it work.
I'll try to investigate this more.

mans0n replied:
So this seems to work fine.
```
	mdio-bus {
		status = "okay";

		ethernet-phy@0 {
			reg = <0>;
		};

		ethernet-phy@1f {
			reg = <0x1f>;
			phy-mode = "rgmii";
		};
```
But I don't feel that's much different from the current state... I don't even understand why I need phy0 here.

musashino205 replied:
WN-AC1167GR had the same issue when I worked for supporting.

> It seems that our eth driver will set up external mt7530 automatically with only the 0x1f node.

I read the driver source and noticed it when that work, and I also deleted phy0 - phy4. But only port0 became unable to communicate. So, for the time being, I left all phys.

musashino205 commented:
@mans0n Confirmed, WN-AC1167GR has that problem. And WN-AC733GR3 has the same.
In these models, it seems that the switch needs to be pre-initialized.
(I didn't notice that script when I worked for adding support of these models...)

However, WN-AC733GR3 has a RTL8367RB switch and the driver is not loaded when entering failsafe-mode. So RTL8367RB wouldn't be detected by the kernel in failsafe-mode and cannot be pre-initialized.

mans0n commented:
@musashino205 Thanks for your confirmation. So mt7620 devices having external ethernet switch seem to have this problem.

> the driver is not loaded when entering failsafe-mode.

Perhaps we should add modprobe or something in ramips_switchdev_from_board()...

mans0n commented:
@musashino205 If you have time, could you please review this PR (and possibly also #2540)? They're quite similar to f305ce5c3577 so it won't take much time.
@adrianschmutzler Do you think this is a good place to add a preinit switch patch for WN-AC733GR3?

mans0n committed:
ramips: increase spi-max-frequency for ipTIME mt7620 devices

This commit increases the hardware SPI frequency from 24.2MHz to 48.3MHz.

[    5.314163] m25p80 spi0.0: speed: 24166666/40000000, rate: 8, prescal: 2, loops: 226
[    5.076323] m25p80 spi0.0: speed: 48333333/50000000, rate: 4, prescal: 1, loops: 162

`time cat /dev/mtd2 >/dev/null` is reduced from 5.64s to 4.36s on A104ns,
and from 11.39s to 8.81s on A1004ns.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
Removed preinit patch due to a8d62a4eb1ce.
The commit really resolved the failsafe ethernet problem. I can't just believe how it fixed the bug so easily... xD

And also added a spi frequency patch.

981213 commented:
Merged into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git
Thanks!

mans0n commented:
@981213 
Thanks for the merge! :)
I should have rebased the patch and tested v5.4 earlier, I was too busy, sorry. I added a fix patch here. mans0n/openwrt@986c45a
And could you please review #2540 as well? It's very similar to A1004ns so it won't take much time for you to review.

981213 commented:
oops

fix pushed.
